core/node/eth_watch/src/event_processors/gateway_migration.rs

```diff
@@ -0,0 +1,67 @@
+use anyhow::Context as _;
+use zksync_dal::{eth_watcher_dal::EventType, Connection, Core, CoreDal};
+use zksync_types::{api::Log, L1BlockNumber, L2ChainId, H256, U256};
+
+use crate::event_processors::{EventProcessor, EventProcessorError, EventsSource};
+
+#[derive(Debug)]
+pub struct GatewayMigrationProcessor {
+    l2chain_id: L2ChainId,
+}
+
+impl GatewayMigrationProcessor {
+    pub fn new(l2chain_id: L2ChainId) -> Self {
+        Self { l2chain_id }
+    }
+}
+
+#[async_trait::async_trait]
+impl EventProcessor for GatewayMigrationProcessor {
+    async fn process_events(
+        &mut self,
+        storage: &mut Connection<'_, Core>,
+        events: Vec<Log>,
+    ) -> Result<usize, EventProcessorError> {
+        for event in &events {
+            let chain_id = U256::from_big_endian(
+                event
+                    .topics
+                    .get(1)
+                    .copied()
+                    .context("missing topic 1")?
+                    .as_bytes(),
+            );
+
+            if L2ChainId::from(chain_id.as_u32()) != self.l2chain_id {
+                continue;
+            }
+
+            storage
+                .server_notifications_dal()
+                .save_notification(
```

Comment:
Do we care about atomicity here? I.e., would saving all events in a transaction / single DB query make a difference? (Off-topic: Currently it doesn't look like `EthWatch::loop_iteration()` calls `process_events()` for processors and updates the next block to process for the processor in a transaction, which may lead to a process the same events multiple times; is this OK?)

Reply:
It's ok, until the order is correct.

Reply:
If you say so 🙃